src/libraries/System.Text.Json/gen/Reflection/TypeExtensions.cs

```diff
@@ -109,6 +109,30 @@ public static bool IsNullableValueType(this Type type, out Type? underlyingType)
             return false;
         }
 
+        public static bool CanContainNullableReferenceTypeAnnotations(this Type type)
+        {
+            // Returns true iff Type instance has potential for receiving nullable reference type annotations,
+            // i.e. the type is a reference type or contains generic parameters that are reference types.
+
+            if (!type.IsValueType)
+            {
+                return true;
+            }
+
+            if (type.IsGenericType)
+            {
+                foreach (var genericParam in type.GetGenericArguments())
```

Comment:
How do generic arguments impact the nullability of the actual variable? E.g. `MyStruct<object>` is non-nullable but this method will return `true`. Maybe I'm not seeing how this generates the usage of `!`.

Also nit\style: `Type` instead of `var`

Reply:
This is meant to catch symbols such as `MyStruct<object?>`. While the outer type is not itself nullable a suppression is still necessary at the serialization callsites.

Under the current `TypeWrapper `approach the source generator is not capable of distinguishing between `MyStruct<object>` and `MyStruct<object?>` symbols (in fact even if we added logic to obtain this information from the underlying `ITypeSymbol`, we are [caching specs up to `System.Type` equality](https://statics.teams.cdn.office.net/evergreen-assets/safelinks/1/atp-safelinks.html)) so we could end up getting inconsistent results.

Instead, this approach checks `Type` for _potential for nullability_. `MyStruct<object>` returns true since it has the same `System.Type` representation as the `MyStruct<object?>` symbol. It can result in false positives though.

Reply:
I see so we are erroring on the side of being nullable, since this code will return true for `MyStruct<object>` and `MyStruct<object?>`.

Perhaps we should try to use the new reflection `System.Reflection.NullabilityInfo` API mentioned in https://github.com/dotnet/runtime/issues/59464#issuecomment-930507192.

Reply:
Can we make `NullabilityInfo` work over `TypeWrapper` instances though? My thinking goes that we should consider removing the `TypeWrapper` infrastructure altogether for 7.0.0, since it's not the most appropriate abstraction layer for source gen.